astropy/modeling/tests/test_core.py

```diff
@@ -984,40 +984,16 @@ def test__validate_input_shape():
 def test__validate_input_shapes():
     model = models.Gaussian1D()
     model._n_models = 2
-    inputs = [mk.MagicMock() for _ in range(3)]
-    argnames = mk.MagicMock()
-    model_set_axis = mk.MagicMock()
-    all_shapes = [mk.MagicMock() for _ in inputs]
 
-    # Successful validation
-    with mk.patch.object(
-        Model, "_validate_input_shape", autospec=True, side_effect=all_shapes
-    ) as mkValidate:
-        with mk.patch.object(core, "check_broadcast", autospec=True) as mkCheck:
-            assert mkCheck.return_value == model._validate_input_shapes(
-                inputs, argnames, model_set_axis
-            )
-            assert mkCheck.call_args_list == [mk.call(*all_shapes)]
-            assert mkValidate.call_args_list == [
-                mk.call(model, _input, idx, argnames, model_set_axis, True)
-                for idx, _input in enumerate(inputs)
-            ]
+    # Full success
+    inputs = [np.array([[1, 2], [3, 4]]), np.array([[5, 6], [7, 8]])]
+    assert (2, 2) == model._validate_input_shapes(inputs, model.inputs, 1)
 
-    # Fail check_broadcast
-    MESSAGE = r"All inputs must have identical shapes or must be scalars"
-    with mk.patch.object(
-        Model, "_validate_input_shape", autospec=True, side_effect=all_shapes
-    ) as mkValidate:
-        with mk.patch.object(
-            core, "check_broadcast", autospec=True, return_value=None
-        ) as mkCheck:
-            with pytest.raises(ValueError, match=MESSAGE):
-                model._validate_input_shapes(inputs, argnames, model_set_axis)
-            assert mkCheck.call_args_list == [mk.call(*all_shapes)]
-            assert mkValidate.call_args_list == [
-                mk.call(model, _input, idx, argnames, model_set_axis, True)
-                for idx, _input in enumerate(inputs)
-            ]
+    # Fails because the input shape of the second input has one more axis which
+    # for which the first input can be broadcasted to
+    inputs = [np.array([[1, 2], [3, 4]]), np.array([[5, 6], [7, 8], [9, 10]])]
+    with pytest.raises(IncompatibleShapeError):
```

Comment:
Changing from ValueError to IncompatibleShapeError is a behavior change too. I don't think we should backport it if you want to keep this change, though from the comment above, I think this is going to be reverted?

Reply:
Right, but the point is we never _did_ raise a ValueError here, because the code that would have re-raised the error was broken (since it was checking for the value None, which check_broadcast never returns).

Don't be fooled by this change to the unit test: it never raised a ValueError. The unit test made it _look like_ it did, because it mocked up the underlying function and the mock did not match the behavior of the real function. So the unit test was wrong too.

Reply:
Okay. Lemme see for myself what the test does now on main. Maybe then I will understand.

Reply:
Okay, yes, you are right, sorry. But what about https://github.com/astropy/astropy/pull/15209/files#r1300237995 ? Does that have to be addressed?

Reply:
I'm not sure what that link points to now.

Reply:
That was @WilliamJamieson 's comment on exception message.